stdlib/public/runtime/Mutex.cpp

```diff
@@ -0,0 +1,139 @@
+//===--- Mutex.cpp - Lockables --------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Mutex, Condition, and Scoped lock abstactions for use in Swift runtime.
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Runtime/Mutex.h"
+
+#include <pthread.h>
+#include <stdlib.h>
+#include <errno.h>
+#include "swift/Runtime/Debug.h"
+
+using namespace swift;
+
+#define reportError(PThreadFunction) \
+{ \
+  int errorcode = PThreadFunction; \
+  if (errorcode != 0) { \
+    fatalError(/* flags = */ 0, \
+               "[%p] '%s' failed with error '%s'(%d)\n", \
+               this, #PThreadFunction, errorName(errorcode), errorcode); \
+  } \
+}
+
+#define returnTrueOrReportError(PThreadFunction, returnFalseOnEBUSY) \
+{ \
+  int errorcode = PThreadFunction; \
+  if (errorcode == 0) return true; \
+  if (returnFalseOnEBUSY && errorcode == EBUSY) return false; \
+  fatalError(/* flags = */ 0, \
+    "[%p] '%s' failed with error '%s'(%d)\n", \
+    this, #PThreadFunction, errorName(errorcode), errorcode); \
+}
+
+static const char *errorName(int errorcode) {
+  switch (errorcode) {
+    case EINVAL:
+      return "EINVAL";
+    case EPERM:
+      return "EPERM";
+    case EDEADLK:
+      return "EDEADLK";
+    case ENOMEM:
+      return "ENOMEM";
+    case EAGAIN:
+      return "EAGAIN";
+    case EBUSY:
+      return "EBUSY";
+    default:
+      return "<unknown>";
+  }
+}
+
+Condition::Condition() : Data(nullptr) {
+  pthread_cond_t *cond =
+    static_cast<pthread_cond_t*>(malloc(sizeof(pthread_cond_t)));
+
+  if (cond == nullptr) {
+    fatalError(/* flags = */ 0, "pthread condition could not be allocated\n");
+  }
+
+  reportError( pthread_cond_init(cond, nullptr) );
+  Data = cond;
+}
+
+Condition::~Condition() {
+  pthread_cond_t *cond = static_cast<pthread_cond_t*>(Data);
+  reportError( pthread_cond_destroy(cond) );
+  free(cond);
+}
+
+void Condition::notifyOne() const {
+  pthread_cond_t *cond = static_cast<pthread_cond_t*>(Data);
+  reportError( pthread_cond_signal(cond) );
+}
+
+void Condition::notifyAll() const {
+  pthread_cond_t *cond = static_cast<pthread_cond_t*>(Data);
+  reportError( pthread_cond_broadcast(cond) );
+}
+
+MutexImplementation::MutexImplementation(bool checked) : Data(nullptr) {
+
+  int kind =
+    ( checked ? PTHREAD_MUTEX_ERRORCHECK : PTHREAD_MUTEX_NORMAL );
+
+  pthread_mutexattr_t attr;
+  pthread_mutex_t *mutex =
+    static_cast<pthread_mutex_t*>(malloc(sizeof(pthread_mutex_t)));
```

Comment:
Well one side effect of inlining I have to lose the generally pervasive use of const &. That may trip up some compiler optimizations I wanted to depend on. It may not be an issue given visibility the compiler had to the code in the header, etc.

I don't have time today to try out the two forms and then compare the output of the optimized unit tests.